main/init/src/mill/init/maven/codegen.scala

```diff
@@ -0,0 +1,353 @@
+package mill.init.maven
+
+import mill.*
+import mill.api.Ctx
+import org.apache.maven.model.{Dependency, Model}
+
+import scala.collection.immutable.SortedSet
+import scala.jdk.CollectionConverters.*
+
+private[maven] object codegen {
+
+  type BuildFile = (os.Source, os.Path)
+
+  def apply(projectRoot: os.Path, baseModule: String, reader: PomReader)(implicit
+      ctx: Ctx
+  ): Seq[BuildFile] = {
+
+    T.log.info("analyzing Maven project ...")
+
+    type ModelNode = (os.Path, Model, BuildPackage)
+    type BuildPackage = String
+
+    // identify Maven models and corresponding Mill packages
+    val (models, modules) = {
+      val (models, modules) = (Seq.newBuilder[ModelNode], Map.newBuilder[ModelId, BuildPackage])
+
+      def loop(pomDir: os.Path, pkg: String): Unit = {
+        val model = reader(pomDir)
+        val modelId = (model.getGroupId, model.getArtifactId, model.getVersion)
+        models += ((pomDir, model, pkg))
+        modules += (modelId -> pkg)
+        model.getModules.asScala.foreach(dir => loop(pomDir / dir, s"$pkg.`$dir`"))
+      }
+
+      loop(projectRoot, "build")
+      (models.result(), modules.result())
+    }
+
+    T.log.info(s"found ${modules.size} module(s)")
+
+    // root model
+    val model0 = models.head._2
+    // identify metadata from root model
+    val (baseModuleSource, testModule) = {
+      val licenses = model0.getLicenses.iterator().asScala.map { license =>
+        val name = Option(license.getName).getOrElse("")
+        val url = Option(license.getUrl).getOrElse("")
+        s"""License("$name", "$name", "$url", false, false, "repo")"""
+      }
+      val developers = model0.getDevelopers.iterator().asScala.map { dev =>
+        s"""Developer("${
+            Option(dev.getId).getOrElse("")
+          }", "${
+            Option(dev.getName).getOrElse("")
+          }", "${
+            Option(dev.getUrl).getOrElse("")
+          }", Option(${
+            Option(dev.getOrganization).fold("null")(escape)
+          }), Option(${
+            Option(dev.getOrganizationUrl).fold("null")(escape)
+          }))"""
+      }
+      val testModule = model0.getDependencies.iterator().asScala.collectFirst {
+        case dep if dep.getGroupId == "junit" => "TestModule.Junit4"
+        case dep if dep.getGroupId == "org.junit.jupiter" => "TestModule.Junit5"
+      }
+      val testFrameworkIvyDeps = testModule.collect {
+        case "TestModule.Junit4" => "ivy\"com.novocode:junit-interface:0.11\""
+        case "TestModule.Junit5" => "ivy\"com.github.sbt.junit:jupiter-interface:0.13.0\""
+      }.toSeq
+      val testModuleSource =
+        s"""  trait Tests extends ${testModule.fold("MavenTests")(mod => s"$mod with MavenTests")} {
+           |
+           |    override def defaultCommandName() = "test"
+           |${
```

Comment:
As a first approximation, we should never have nested strings inside `${}` interpolation block, and `${}` blocks should never be multi-line. If they are getting complex, their contents should be pulled out into a named `val` earlier